controllers/drain_controller.go

```diff
@@ -0,0 +1,501 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	sriovnetworkv1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	constants "github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/consts"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/drain"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/platforms"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/utils"
+	"github.com/k8snetworkplumbingwg/sriov-network-operator/pkg/vars"
+)
+
+type DrainReconcile struct {
+	client.Client
+	Scheme   *runtime.Scheme
+	recorder record.EventRecorder
+	drainer  drain.DrainInterface
+
+	nodesInReconcile      map[string]interface{}
+	nodesInReconcileMutex sync.Mutex
+	drainCheckMutex       sync.Mutex
+}
+
+func NewDrainReconcileController(client client.Client, Scheme *runtime.Scheme, recorder record.EventRecorder, platformHelper platforms.Interface) (*DrainReconcile, error) {
+	drainer, err := drain.NewDrainer(platformHelper)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DrainReconcile{
+		client,
+		Scheme,
+		recorder,
+		drainer,
+		map[string]interface{}{},
+		sync.Mutex{},
+		sync.Mutex{}}, nil
+}
+func (dr *DrainReconcile) tryLockNode(nodeName string) bool {
+	dr.nodesInReconcileMutex.Lock()
+	defer dr.nodesInReconcileMutex.Unlock()
+
+	_, exist := dr.nodesInReconcile[nodeName]
+	if exist {
+		return false
+	}
+
+	dr.nodesInReconcile[nodeName] = nil
+	return true
+}
+
+func (dr *DrainReconcile) unlockNode(nodeName string) {
+	dr.nodesInReconcileMutex.Lock()
+	defer dr.nodesInReconcileMutex.Unlock()
+
+	delete(dr.nodesInReconcile, nodeName)
+}
+
+//+kubebuilder:rbac:groups="",resources=nodes,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=sriovnetwork.openshift.io,resources=sriovnodestates,verbs=get;list;watch
+//+kubebuilder:rbac:groups=core,resources=events,verbs=create;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (dr *DrainReconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	// try to lock the node to this reconcile loop.
+	// if we are not able this means there is another loop already handling this node, so we just exist
+	if !dr.tryLockNode(req.Name) {
```

Comment:
I think that implementation of the queue in the client-go package [1] will protect you from this case. You will have only 1 reconcile request for one key(node name in our case) even if max parallel reconciles is 50.

[1] https://github.com/adrianchiris/sriov-network-operator/blob/56530b9c0e1c893d686f545df6876ddce807950f/vendor/k8s.io/client-go/util/workqueue/queue.go#L163

Reply:
done thanks!